libslide/src/parser.rs

```diff
@@ -0,0 +1,135 @@
+use crate::scanner::types::Token;
+use crate::scanner::types::TokenType;
+
+pub enum Expr {
+    Float(f64),
+    Int(i64),
+    BinOp(BinOp),
+    // I added un op even though I havent implemented + or -
+    UnaryOp(UnaryOp),
+}
+
+pub struct BinOp {
+    pub item: Token,
+    pub lhs: Box<Expr>,
+    pub rhs: Box<Expr>,
+}
+
+pub struct UnaryOp{
+    pub item: Token, 
+    pub rhs: Box<Expr>,
+}
+
+pub struct Parser<'a>{
+    input: &'a Vec<Token>,
+    index: usize,
+    cur_token: Token,
```

Comment:
If we have an index, why do we need a `cur_token`? If the worry is with borrowing, let's just own the input instead (`Vec<Token>`)

Reply:
the reason we have an index is to hold a variable to iterate through the vector and the cur_token. The reason we have a cur token is so that we don't need to keep passing the token from each recursive call since they all have access to self. This could defniitely be reworked i don't disagree